src/backend/PortabilityService.ConfigurationService/Controllers/ConfigurationController.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using ConfigurationControllerResources = PortabilityService.ConfigurationService.Resources.Controllers.ConfigurationController;
+using Microsoft.AspNetCore.Hosting;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.Extensions.Configuration;
+using Microsoft.Extensions.Localization;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using System;
+using System.Linq;
+
+namespace PortabilityService.ConfigurationService.Controllers
+{
+    [Route("api/[controller]")]
+    public class ConfigurationController : Controller
+    {
+        private readonly IConfiguration _configuration;
+        private readonly IHostingEnvironment _hostingEnvironment;
+        private readonly IStringLocalizer<ConfigurationController> _localizer;
+        private readonly ILogger _logger;
+        private readonly string _baseConfigSectionName;
+
+        public ConfigurationController(IConfiguration configuration,
+                                       IHostingEnvironment hostingEnvironment,
+                                       IStringLocalizer<ConfigurationController> localizer,
+                                       ILogger<ConfigurationController> logger = null,
+                                       string baseConfigSectionName = "PortabilityServiceSettings")
+        {
+            _configuration = configuration ?? throw new ArgumentNullException(nameof(configuration));
+            _hostingEnvironment = hostingEnvironment ?? throw new ArgumentNullException(nameof(hostingEnvironment));
+            _localizer = localizer ?? throw new ArgumentNullException(nameof(localizer));
+            _logger = logger ?? new NullLogger<ConfigurationController>();
+
+            if (string.IsNullOrEmpty(baseConfigSectionName))
+            {
+                throw new ArgumentNullException(nameof(baseConfigSectionName));
+            }
+            else
+            {
+                _baseConfigSectionName = baseConfigSectionName;
+            }
+        }
+
+        [HttpGet]
+        public ObjectResult Get()
+        {
+            _logger.LogInformation(BuildLogMessage(nameof(Get), _localizer[nameof(ConfigurationControllerResources.ReturningAllConfigSettings)].Value));
+            return Ok(_configuration.GetSection(_baseConfigSectionName).AsEnumerable());
+        }
+
+        [HttpGet("section/{sectionName}")]
+        public ObjectResult GetSection(string sectionName)
+        {
+            if (string.IsNullOrEmpty(sectionName))
+            {
+                var errorMessage = _localizer[nameof(ConfigurationControllerResources.SectionNameNullOrEmpty)].Value;
+                _logger.LogError(BuildLogMessage(nameof(GetSection), errorMessage));
```

Comment:
Passing parameters for diagnostic messages separately (instead of formatting them into the message string) is preferable for later searching in structure logging output.

So, instead of `_logger.LogError($"{nameof(GetSection)}: {errorMessage}")`, it would be better to have `_logger.LogError("{ActionName}: {ErrorMessage}", nameof(GetSection), errorMessage)`. If we do it that way, then if we log to some sort of structured-logging output (App Insights, for example), we'll be able to construct queries with clauses like "if (ActionName == 'GetSection')" or something like that.

Reply:
Changed